cpp/src/c_api/bfs.cpp

```diff
@@ -16,54 +16,220 @@
+            typename weight_t,
+            bool store_transposed,
+            bool multi_gpu>
+  void operator()()
+  {
+    // FIXME: Think about how to handle SG vice MG
+    if constexpr (!cugraph::is_candidate<vertex_t, edge_t, weight_t>::value) {
+      unsupported();
+    } else {
+      // BFS expects store_transposed == false
+      if constexpr (store_transposed) {
+        error_code_ =
+          cugraph::c_api::transpose<vertex_t, edge_t, weight_t, store_transposed, multi_gpu>(
+            handle_, graph_, error_.get());
+        if (error_code_ != CUGRAPH_SUCCESS) return;
+      }
+
+      auto graph =
+        reinterpret_cast<cugraph::graph_t<vertex_t, edge_t, weight_t, false, multi_gpu>*>(
+          graph_->graph_);
+
+      auto graph_view = graph->view();
+
+      auto number_map = reinterpret_cast<rmm::device_uvector<vertex_t>*>(graph_->number_map_);
+
+      rmm::device_uvector<vertex_t> vertex_ids(graph->get_number_of_vertices(),
+                                               handle_.get_stream());
+      rmm::device_uvector<vertex_t> distances(graph->get_number_of_vertices(),
+                                              handle_.get_stream());
+      rmm::device_uvector<vertex_t> predecessors(0, handle_.get_stream());
+
+      if (compute_predecessors_) {
+        predecessors.resize(graph->get_number_of_vertices(), handle_.get_stream());
+      }
+
+      //
+      // Need to renumber sources
+      //
+      renumber_ext_vertices<vertex_t, multi_gpu>(handle_,
+                                                 sources_->as_type<vertex_t>(),
+                                                 sources_->size_,
+                                                 number_map->data(),
+                                                 graph_view.get_local_vertex_first(),
+                                                 graph_view.get_local_vertex_last(),
+                                                 do_expensive_check_);
+
+      cugraph::bfs<vertex_t, edge_t, weight_t, multi_gpu>(
+        handle_,
+        graph_view,
+        distances.data(),
+        compute_predecessors_ ? predecessors.data() : nullptr,
+        sources_->as_type<vertex_t>(),
+        sources_->size_,
+        direction_optimizing_,
+        static_cast<vertex_t>(depth_limit_),
+        do_expensive_check_);
+
+      cugraph::detail::sequence_fill(
+        handle_.get_stream(), vertex_ids.data(), vertex_ids.size(), vertex_t{0});
+
+      std::vector<vertex_t> vertex_partition_lasts = graph_view.get_vertex_partition_lasts();
+
+      unrenumber_int_vertices<vertex_t, multi_gpu>(handle_,
+                                                   vertex_ids.data(),
+                                                   vertex_ids.size(),
+                                                   number_map->data(),
+                                                   vertex_partition_lasts,
+                                                   do_expensive_check_);
```

Comment:
Or we may even skip copying if we can return a const pointer.

Reply:
Good point.

The `number_map` is part of the `cugraph_graph_t`.  It will be available until either the `cugraph_graph_t` is freed, or until the next time the graph is transposed.

Safest would be to copy, I think.  Avoiding the copy would require the user to be in the habit of completely consuming the results of one algorithm before beginning the next.  This would certainly be a good practice - especially considering GPU memory is frequently the limiting resource.  But I'm not sure requiring it is ideal.  Thoughts?

Reply:
Yeah... and I guess we'd better copy, but then, the following question is to copy when (or where?) If non-python users may be OK (especially hard-core HPC C users) with the additional constraint (number_map->data() should be remain valid till the vertex data get consumed), we may copy in the python side (I assume worrying about when to free resources is against the spirit of most python users).

Reply:
I added the copy on the C side in the latest push.

I'll add an issue to grapple with this longer term issue.  Perhaps we could use an std::shared_ptr in the opaque object.  This would allow the `cugraph_graph_t` and the `cugraph_pagerank_result_t` (and other algorithms results) to share the number_map.  When the graph gets transposed any existing copies would still point to the original mapping but the graph would point to the new number map.

The copy works for now, this would be a memory optimization we can address later.

Reply:
Created #1947